### ISSUE TITLE
Segments signedness issue

### ISSUE THREAD

**Issue by fabianfreyer** [Vector35/binaryninja-api#2881]
**Version and Platform (required):**
- Binary Ninja Version: 3.0.3222-dev (2022-01-26 06:24:12)
- OS: macOS
- OS Version: 11

**Bug Description:**
Segment addresses and lengths seem to have a signedness issue.

**Steps To Reproduce:**
1. Create a small test binary:
   ```bash
   truncate -s 4 empty
   ```
2. open with binja, select any arch.
3. Run the following commands in the python shell:
    ```python
    # can change these lengths arbitrarily within size
    len1 = 1
    len2 = 2
    bv.add_user_segment(0, len1, 0, len1, 5)
    bv.add_user_segment(len1, len2, len1, len2, 7)
    ```
    Segments now look like this:
    ```
    Type: Mapped
    Platform: x86_64
    Architecture: x86_64
    
    Segments:
    r-x  0x00000000-0x00000001 
    rwx  0x00000001-0x00000003 
    r-x  0x00000003-0x00000004 
    ```
3. Save bndb
4. Open bndb
    Segments now look like this:
    ```
    Type: Mapped
    Platform: x86_64
    Architecture: x86_64
    
    Segments:
    r-x  0x00000000-0x00000004 
    r-x  0xfffffffffffffffd-0xfffffffffffffffe 
    rwx  0xfffffffffffffffe-0x00000000 
    ```

**Expected Behavior:**
Segments should be the same before and after saving.

**Screenshots:**
Before saving:
![image](https://user-images.githubusercontent.com/3423837/151235635-3b0be89e-1fe7-4fa6-a681-c370d04beebf.png)

After saving:
![image](https://user-images.githubusercontent.com/3423837/151235697-05390570-1e2f-445a-b941-bb27a0f21855.png)

**Comment by fabianfreyer** [comment]
Closing in favor of #2882

**Comment by psifertex** [comment]
#2882 has been renamed to describe a more broad feature that would include this fix as a super set but has been pushed out further to 3.2 so we're opening this to track just the existing bug and potentially a quicker temporary fix in the core during creation of user segments when no auto-segments exist such that this issue no longer occurs.

**Comment by psifertex** [comment]
Also, note that we believe that using the open-with-options JSON specification for segments is a temporary work-around in the meantime.

**Comment by bpotchik** [comment]
Previously, normalized offsets were based on `bv.start`. Now normalized offsets are based on `bv.image_base`. This allows us to support adding 'user' segments and/or memory regions before the image base. Now, rebasing and save/load operations preserve the memory map correctly.

Fixed in `4.2.5758`.
